File: typical90/src/bin/001.rs
use proconio::input;
use std::vec::Vec;

fn solve(n: i32, l: i32, k: i32, a: &Vec<i32>) -> i32 {
    // 二分探索
    let mut ok: i32 = 0;
    let mut ng: i32 = l;
    while (ok - ng).abs() > 1 {
        let mid = (ok + ng) / 2;
        // mid 以上の長さのピースに分割できるか
        if is_ok(mid, n, l, k, a) {
            ok = mid;
        } else {
            ng = mid;
        }
    }
    ok
}

fn is_ok(score: i32, n: i32, l: i32, k: i32, a: &Vec<i32>) -> bool {
    let mut cnt = 0;
    let mut pre = 0;
    for i in 0..n {
        // score 以上の長さのピースに分割する
        if a[i as usize] - pre >= score {
            cnt += 1;
            pre = a[i as usize];
        }
    }
    // k 個以上に分割できるか
    if cnt > k || (cnt == k && l - pre >= score) {
        return true;
    }
    return false;
}

fn main() {
    input! {
        n: usize,
        l: usize,
        k: usize,
        a: [i32; n],
    }
    println!("{}", solve(n as i32, l as i32, k as i32, &a));
}

File: typical90/py/001.py
from os import getenv


def solve(N, L, K, A):
    """
    >>> solve(3, 34, 1, [8, 13, 26])
    13
    >>> solve(7, 45, 2, [7, 11, 16, 20, 28, 34, 38])
    12
    """

    def is_ok(x):
        count = 0
        last = 0
        for a in A:
            if a - last >= x:
                count += 1
                last = a
        if L - last >= x:
            count += 1
        return count >= K + 1

    return binary_search_max(0, L + 1, is_ok)


def binary_search_max(ok, ng, is_ok):
    while abs(ok - ng) > 1:
        mid = (ok + ng) // 2
        if is_ok(mid):
            ok = mid
        else:
            ng = mid
    return ok


if getenv("DOCTEST"):
    import doctest

    doctest.testmod()

if __name__ == "__main__" and not getenv("DOCTEST"):
    N, L = map(int, input().split())
    K = int(input())
    A = list(map(int, input().split()))

    print(solve(N, L, K, A))
